Add hypertext support to tooltips

Wuzzy2 opened the pull request:
### Summary

Adds fancy hypertext support to tooltips to make them look pretty.

![supertip](https://github.com/user-attachments/assets/d92171a5-3b72-4993-98bf-546877f6ba13)

### Background of this PR

I was asked by @kilbith to submit this PR, who did the work for the first commit. @kilbith is unable to do it alone because of a ban. I have looked at the new feature and tested it myself, checked the code and it looks OK. I believe it will be like a very good addition to Minetest, but it needs work. Please read the code carefully yourself before running it.

I do not know if our policies allow a PR to be made like this. If you dislike what I try to do here, I feel very sorry. Just close this PR and it's over. But if this is OK:

We have explicit, written permission to use this work under LGPLv2.1+. Quote from e-mail conversation with @kilbith:

> 1. Permission is expressly granted by my organization to license this code
> under the GNU LGPL v2.1+ terms.

This e-mail has been cc'ed to @celeron55 as well.

~~**IMPORTANT: DO NOT SQUASH!** The first commit MUST be left unaltered, which contains the requested author identity. I was given the patch only under this condition, to which I agreed. The agreement allows to *add* new commits to build on the initial commit tho. So bugfixes and changes all have to go in NEW commits.~~ (kilbith has retracted the no-squash requirement, so it no longer applies)

### Feature description

This feature will add hypertext support for tooltips by introducing a new `hypertip` tooltip element. This element works like `tooltip` but you can write hypertext (like for the `hypertext` element).

There are two modes: dynamic and static. Dynamic supertips will show up next to the cursor, just like normal ones. Static supertips appear at a fixed formspec coordinate.

Also, hypertext tooltips can be *styled*.

See the `lua_api.md` for details.

### How to test

There is an example mod  with an example formspec under <https://codeberg.org/Wuzzy/temporary_luanti_hypertip_test_mod> that you can install and try (see instructions over there). This URL is *temporary* and will be deleted once this PR is merged.

This PR itself also extends the `/test_formspec` command in DevTest by adding a "Tooltip" tab. This new tab demonstrates a few minimalistic normal tooltips and hypertext tooltips.

### Demo

Here is a video demonstration:

<https://www.youtube.com/watch?v=RAASW_Rj8eQ>

### Feedback

Feedback of the concept is greatly appreciated. Especially please comment on the syntax and design choices.

Also report any bugs I didn't find myself before.

nerzhul commented:
it sounds to be a very good enhancement.
A question about this, cannot we just have our current tip code using this new parsing method ?
Is supertip the good name ? 
If it's a new UI addition, don't we need a protocol bump ? what happen if a server uses it and the client is older ?

nerzhul commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3797,6 +3884,39 @@ void GUIFormSpecMenu::showTooltip(const std::wstring &text,
+	if (spec.floating) {
+		/* Issue with floating tooltips' positioning here.
+		   Set hardcoded values that only works on 16:10/4K displays. */
+		X = m_pointer.X - 1120;
```
what should be do with this ?

nerzhul commented:
another comment, can we have multiple examples in devtest ?

Wuzzy2 commented:
Well, all these things are open for debate. I haven't deeply analyzed the code myself yet, only read it twice. The only thing that matters is that the 1st commit remains intact.

I agree that `supertip` might not be the ideal name, but I'm neutral about this. I don't know why @kilbith chose to add a new formspec element rather than extending `tooltip`. But I can think of some reasons myself: Keeping the default tooltip simple and for compability. Also, please keep in mind that the `supertip` syntax will parse hypertext, so `<` and `>` have a special meaning. While the default `tooltip` does not give them special treatment. We cannot make `tooltip` suddenly parse hypertext, it will break tooltips that already use `<` and `>` and other special characters as plain text.

Although a new param could be added to toggle hypertext parsing. And it is off by default.

supertip does other things as well: It is also stylable and allows absolute positions.

But yeah, thinking about it, it MAY be possible to extend the `tooltip` element instead, the hard part is to make sure it'll all be backwards-compatible and not making the new syntax too messy. Annoyingly, for this to work, I'm afraid we'd have to add the new parameters *after* the text. But as long it works …

Another open question is whether this will require a new formspec version. I don't the rules for that.

As I understand the code, it does not currently handle the case of an old client receiving a supertip. This was not tested yet, so assume the worst. :D

Anyway. I think the top priority for me is to fix all the bugs first to get it to work properly. Optimization/cleanup can come later.

Wuzzy2 commented:
There *are* multiple exampels in DevTest, in the `/test_formspec` command. Or did you mean the current examples are not enough? If so, I agree. Adding more DevTest examples is on my TODO list. Currently missing is an example for styling.

About this code:

> /* Issue with floating tooltips' positioning here.
> Set hardcoded values that only works on 16:10/4K displays. */

@kilbith told me this is a workaround for a known bug, to "force" the feature to work *at all*. The bug here is that the supertip is not positioned correctly if it's dynamic. It is often placed far away from the cursor, changing with screen size. So what needs to be done is to fix the bug, making sure dynamic supertips appear correctly at the cursor. (dynamic supertip = tooltip that appears relative to the cursor, like the default tooltip)

nerzhul commented:
for the devtest part, ignore my comment, i just skipped the file in my brain, it's ok :D

Wuzzy2 committed:
Fix positioning of dynamic supertips

Wuzzy2 committed:
Supertip: Make static position argument mandatory

Wuzzy2 committed:
Fix segfault if invalid arg in supertip element

Wuzzy2 commented:
Progress report: Dynamic supertips (i.e. supertips that follow the cursor like normal tooltips) now have the correct positioning. I also changed the syntax. The `posX,posY` argument is now called `staticPos` and is mandatory, meaning the element now has a fixed argument count. This should make things clearer and easier to understand. I also fixed a segfault when this argument had invalid syntax.

2 known bugs remain (see above).

> for the exteneded parsing, we can imagine extended=true attribute in tooltip directly, switching the parsing mode maybe ?

Yeah, I was thinking about something like thit. I could probably work, I just need to figure out the proper syntax while maintaining compat.

grorp reviewed (commented):
Thoughts:

- "hypertext + tooltip = supertip"? Not rather "hypertip"?

- I feel like having a new element for hypertext tooltips, instead of reusing the existing tooltip element by adding new parameters at the end, is less confusing. There's also the `fontcolor` and `bgcolor` parameters for regular tooltips, which wouldn't make much sense for hypertext tooltips.

- Compared to regular tooltips, hypertext tooltips seem to miss the feature where you can "attach" them to formspec elements. It might make sense to support this for consistency.

- Not required in this PR, but I think it would be great if item (metadata) tooltips could also optionally use hypertext.

rubenwardy commented:
I recommend the obvious name: `tooltip_hypertext`

Commit:
Fix supertip flicker; enable supertip border

Wuzzy2 committed:
Document supertip border default as true

Commit:
Allow to bind supertip to a formspec element

Wuzzy2 committed:
Clean up supertip documentation

Wuzzy2 committed:
DevTest: Add tests for element-bound tooltips

nerzhul commented on src/gui/guiHyperText.cpp:
```diff
@@ -1187,8 +1236,11 @@ void GUIHyperText::draw()
-	m_drawer.draw(AbsoluteClippingRect,
-			m_display_text_rect.UpperLeftCorner + m_text_scrollpos);
+	if (m_drawer_ready)
+		m_drawer.draw(AbsoluteClippingRect,
```
please add braces, it's a very dangerous 2 lines no braces notation

nerzhul commented on src/gui/guiHyperText.h:
```diff
@@ -177,6 +179,8 @@ class TextDrawer
 	void draw(const core::rect<s32> &clip_rect,
 			const core::position2d<s32> &dest_offset);
+	void drawBackgroundImage(video::IVideoDriver *driver, const ParsedText &m_text, const core::rect<s32> &clip_rect);
+	ParsedText &getText() { return m_text; }
```
i don't like this very dangerous pattern, it's better to have a copy and a setting i think. ok it's optimized memory side, but having full access to a member ref sounds bad

Commit:
Fix delay on bound supertips

Wuzzy2 committed:
Coding style fix in guiHyperText.cpp

nerzhul commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1755,6 +1754,118 @@ void GUIFormSpecMenu::parseHyperText(parserData *data, const std::string &elemen
+
+		e->setVisible(false);
+		e->drop();
+	} else {
```
useless else, just return on the previous braces and remove this else braces

nerzhul commented:
appart some style it sounds good, anything remaining to fix to move out of draft ?

Wuzzy2 commented:
The one remaining bug is that the tooltip width is too small and may even make its content disappear when the window is small.
Currently, tooltip width is relative to formspec units but I think this is the wrong approach. I think it is better is to use typographic em units.

I agree the `&getText` is indeed risky and should be replaced; I haven’t gotten a replacement to work so far and haven’t made another try at this.

Wuzzy2 committed:
Tweak indent levels in guiFormSpecMenu.cpp

Wuzzy2 commented:
The reason why there has been no progress recently is because I'm stuck and don't know how to fix the remaining bugs/issues. :-(

Sorry.

Commit:
Supertip: Fix insecure &getText

Wuzzy2 commented:
I'm happy to report that kilbith fixed the `&getText` issue and I decided to move this PR out of draft.

Now the only bug is the tooltip width issue, as explained in <https://github.com/minetest/minetest/pull/14728#issuecomment-2198091438>.

I think this is a serious issue that should not be ignored. Nowhere else does Minetest make the tooltip text arbitrarily disappear. But I think I have a plan …

I will rebase once the final bug was fixed (don't want to rebase an unfinished PR over and over).

Also, I should probably improve the DevTest formspec again. There is no test for the tooltip styling yet.

I haven’t chosen a different name for the "supertip" formspec element yet, but I am willing to accept suggestions if there is an consensus. Please comment. (So far only 2 people commented, and I am tending towards @rubenwardy's `tooltip_hypertext`.)

Wuzzy2 committed:
DevTest: Add tests for supertip styling

Wuzzy2 committed:
Supertips: Specify width as 'em' units

Wuzzy2 committed:
Remove noclip from supertip style documentation

Wuzzy2 commented:
New changes:
* Fixed tooltip text disappearing with small window sizes
* Redefined supertip width from formspec units to 'em' units, relative to the default font at unmodified size (i.e. no `<big>` tag, etc.)
* Removed `noclip` from supertip style documentation. I've discussed with kilbith and we decided it was pointless
* Rebase completed

There are no more known bugs left and the supertips work as intended!

Test the new witdth system by using different font sizes in the Minetest config. It should scale. Note the only thing I redefined is the size of a `width` unit, not the underlying logic.

The reason why I picked 'em' units is because it seems the most obvious. As a reminder, 'em' is equal to font size so it neatly scales with the font. Another upside is that the tooltip no longer cares about the window size. This was what caused the old bug: The tooltip size was dependent on formspec units, which in turn is dependent on window size. This didn't make sense to me, so I changed it.

The width system has one inherent downside: The width is not flexible nor automatic. The modder MUST pick a width that kinda leaves enough room. Very generous widths will leave a lot of empty space that isn't truncated. A too small width will cut off text, but it is now the job of the modder to figure this one out. This will also become an issue for translated strings, which vary in length. This isn't a bug, but it may be tricky to find a good width.

So for now, I’d say the recommended use of supertips is generally be more generous with width, just in case.

ANYWAY. The system works as intented tho and is ready for review. And thinking about this, I think we're pretty much *forced* to have to specify a width like that anyway. Because hypertext includes right-justified text and it can't know where the "right-most side" is on its own. It would be great if width could be fully automatically calculated, but I think this is stuff for a new PR (if it is even possible or realistic).

In short, in my opinion this PR is now OFFICIALLY ready for review. All the known bugs are fixed.

Wuzzy2 commented:
Oh, one hint for practical usage: If you want a supertip with only one word or line, I recommend to provide more width than the English word (about 1.5×, for the translations) *plus* use `<center>` so the empty space is equally distributed and it looks nicer.

Wuzzy2 commented:
Reminder: I rebased, so the label can be removed (for now).

jingkaimori commented on src/gui/guiHyperText.h:
```diff
@@ -177,6 +179,8 @@ class TextDrawer
 	inline s32 getHeight() { return m_height; };
 	void draw(const core::rect<s32> &clip_rect,
 			const core::position2d<s32> &dest_offset);
+	void drawBackgroundImage(video::IVideoDriver *driver, const ParsedText &m_text, const core::rect<s32> &clip_rect);
```
Why passes `ParserdText` as argument in this method? `m_text` can be obtained from instance.

Wuzzy2 committed:
Remove redundant parameter in guiHyperText

Wuzzy2 commented:
Rebased again.

> Why passes ParserdText as argument in this method? m_text can be obtained from instance.

You're right, this argument is not neccessary, so I removed it. It still works.

Wuzzy2 commented:
Bump.

Wuzzy2 committed:
Fix PointerType in guiFormSpecMenu

Wuzzy2 commented:
I just made a fresh new rebase to account for the changes of the last month so it is ready for testing again.

Please someone review this sometime. Everything that has been requested from us has been done and all known bugs are fixed. This PR has already missed several Luanti releases, I wish this nice feature can make it into the next one.

Wuzzy2 committed:
Add back missing StyleSpec.h include

kilbith commented on src/gui/guiHyperText.cpp:
```diff
@@ -939,6 +957,20 @@ void TextDrawer::draw(const core::rect<s32> &clip_rect,
+		const auto &UpperLeft = clip_rect.UpperLeftCorner;
+		const auto &LowerRight = clip_rect.LowerRightCorner;
+
+		driver->draw2DLine(UpperLeft, core::position2di(LowerRight.X, UpperLeft.Y), color);
```
You probably meant [draw2DRectangleOutline](https://irrlicht.sourceforge.io/docu/classirr_1_1video_1_1_i_video_driver.html#a65efe36e19f0570988848175a8af7bd1) which was removed from IrrlichtMT.

And `draw2DRectangleOutline` was drawing 4x `2DLine` internally.

Wuzzy2 committed:
Clean up some hypertip-related dependencies

Wuzzy2 committed:
Use shared function for tooltip positioning

Wuzzy2 commented:
So most comments are addressed now. I will not do a squash for now until the PR is deemed acceptable. I hope to address the final open comment soon(ish).

(My philosophy is, first get this thing in an acceptable state, THEN do whatever commit cleanup is neccessary. Not the other way around.)

SmallJoker reviewed (commented):
Works OK... with caveats highlighted in the comments.

Wuzzy2 committed:
Hypertip: Clean up the syntax checks

Wuzzy2 committed:
Fix hypertip poorly interacting with screen bounds

Wuzzy2 committed:
Minor refactor in guiFormSpecMenu.cpp

Wuzzy2 commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1762,6 +1761,125 @@ void GUIFormSpecMenu::parseHyperText(parserData *data, const std::string &elemen
+		errorstream << "Invalid hypertip element(" << parts.size() << "): '"
+				<< element << "'"  << std::endl;
+		return;
+	}
```
I already applied the suggested change. If that's enough to satisfy the compatibility requirements, I think this one can be marked as resolved then. Oh, and I just applied the snake case.

Wuzzy2 commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -3683,6 +3859,50 @@ void GUIFormSpecMenu::showTooltip(const std::wstring &text,
+{
+	// Hypertip size and offset
+	s32 tooltip_width = spec.width;
+	s32 tooltip_height = e->getTextHeight() + 5;
```
I personally prefer to use no offset at all and it looks OK in my tests. And IMO with offset it looks a little weird, especially on larger resolution.
But if an offset is desired, I can add a small fixed pixel offset.

For testing, you could make a test formspec where a hypertip activates on a very large rectangle area to ease testing near the boundaries.

SmallJoker reviewed (approved):
Works well enough.

Wuzzy2 committed:
Fix 1-tick long flicker of some hypertips

Wuzzy2 commented:
@nerzhul: GitHub claims you are still "requesting changes" but I think everything is already addressed?

Anyway, just fixed a small "flicker" bug and I guess this PR is fully ready again.

Wuzzy2 committed:
Add "supertip" formspec element

Wuzzy2 committed:
DevTest: Add test for supertip formspec element

Wuzzy2 committed:
Rename 'supertip' to 'hypertip'

appgurueu commented on src/gui/guiHyperText.cpp:
```diff
@@ -1170,8 +1225,14 @@ void GUIHyperText::draw()
+	// We needs 2 iterations for the element positions and total height
+	// to be pre-calculated correctly, so we need to delay the actual drawing
+	// by this many iterations
+	if (m_draw_state >= 2) {
```
This feels like one hell of a hack, with potential negative implications (like noticeable jank: i assume this translates to 2 frames of delay?)

Is there no decently simple way to figure out what this depends on, and make sure the drawing proceeds in such an order that dependencies are fulfilled? e.g., do the drawing in stages, where hypertips are drawn last?

kilbith replied:
IIRC (it's been a long time since I've touched this code), it's depending on `TextDrawer::place` not being ready to determine the correct text dimensions to draw the tooltip with. Note the delay is the 2 first frames *ever* after the tooltip is requested to draw. It's not even remotely noticeable by the user.

Wuzzy2 committed:
Fix typo and style in hypertip DevTest example

Wuzzy2 committed:
Hypertip: Improve code style of argument parsing

Wuzzy2 commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -1763,6 +1762,130 @@ void GUIFormSpecMenu::parseHyperText(parserData *data, const std::string &elemen
+	// size as the first 2 arguments, false otherwise
+	const bool rect_mode = parts[0].find(',') != std::string::npos;
+	// Number of expected hypertip arguments
+	const size_t base_size = rect_mode ? 6 : 5;
```
Done.

kilbith commented on src/gui/guiFormSpecMenu.cpp:
```diff
@@ -2398,7 +2542,10 @@ void GUIFormSpecMenu::parseTooltip(parserData* data, const std::string &element)
-	std::string text = unescape_string(parts[rect_mode ? 2 : 1]);
+	// 1 or 2
+	std::string text = unescape_string(parts[arg_cursor]);
+	arg_cursor++;
```
All of this is superfluous: 

```C++
foo = bar[idx];
idx++;
```

You can just: `foo = bar[idx++];` since incrementation happens after assignment.

Wuzzy2 committed:
Hypertips now respect default tooltip [bg]color

Wuzzy2 committed:
Clean up DevTest tooltip test a little

Wuzzy2 committed:
DevTest: Redo tooltip test, add listcolors

Wuzzy2 commented on src/gui/guiHyperText.cpp:
```diff
@@ -1003,6 +1035,20 @@ void TextDrawer::draw(const core::rect<s32> &clip_rect,
+
+	if (m_text.background_type != m_text.BackgroundType::BACKGROUND_COLOR) {
+		m_text.background_type = m_text.BackgroundType::BACKGROUND_COLOR;
+		m_text.background_color = style.getColor(StyleSpec::BGCOLOR, video::SColor(255,110,130,60));
```
Yes, this hardcoding was bad. I fixed it. Tooltips now take the "global" tooltip color by default, including the one set by `listcolors[]`.

Wuzzy2 committed:
Say that action elem isnt clickable in hypertip

Wuzzy2 committed:
DevTest: Bump tooltip test formspec_version to 9

because it includes an area label

appgurueu commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -149,6 +151,36 @@ class GUIFormSpecMenu : public GUIModalMenu
 
+	struct HyperTipSpec
+	{
+		HyperTipSpec() = default;
```
I think both these constructors can be dropped. For the second constructor, just use aggregate initialization (`HyperTipSpec{...}` instead of `HyperTipSpec(...)`, might want to default-initialize the primitive fields so they can't go uninitialized). This is more fitting for such a data struct.

Wuzzy2 replied:
I’m not too happy with such a change, all the other structs in that file use explicit notation, so this suggestion would break the consistency in the file. Also, the official coding style appears to be silent on constructors, so …

I think it’s better to do *general* code-style changes outside of this PR.

appgurueu replied:
The reason I suggested it is that seeing the changes in this PR (and how I expect the struct to be used in the future), I know that my suggestion results in neater code overall. [^init]

But fair enough, I hadn't examined the surrounding style. It's fine to leave that consistent for now.

I will propose some changes to our style guide eventually, it is silent on too many things.

[^init]: There is a detail with regard to initialization of primitives that don't get default-initialized (e.g. `bool`), a constructor can force their initialization. But the proper solution is simply to give each member a default initializer (e.g. `= false` for `bool`). You have to do this anyways if you want a default constructor that doesn't leave members uninitialized. (The current default constructor does.) Well whatever, off to the style guide rationale it goes...

appgurueu commented on src/gui/guiFormSpecMenu.h:
```diff
@@ -149,6 +151,36 @@ class GUIFormSpecMenu : public GUIModalMenu
+		std::string parent_name;
+		std::string text;
+		core::rect<s32> hover_rect;
+		v2s32 stpos; //< static tooltip position
```
```suggestion
		v2s32 stpos; ///< static tooltip position
```

pretty sure triple `/` is needed for doxygen to pick up on it (https://www.doxygen.nl/manual/docblocks.html)

same below

appgurueu reviewed (commented):
Mostly looking good! Sorry for the late review.

One more minor thing: Shouldn't the margin be on all sides? It looks like the top margin is added to the bottom. This creates an awkward empty space at the bottom.

<img width="163" height="133" alt="Image" src="https://github.com/user-attachments/assets/0ac68a51-e83b-49c8-a879-24d42d6b6f48" />

Wuzzy2 committed:
Hypertips: Remove hardcoded tooltip height offset

Wuzzy2 committed:
Hypertext: Fix margin not applied correctly to top

Wuzzy2 committed:
Fix doxygen markers in guiFormSpecMenu.h

appgurueu reviewed (approved):
Looks correct to me. The `hypertext[]` in `/test_formspec` still renders basically the same as on 5.11 and 5.17-dev, which is with the margin applied on all sides as expected:
<img width="782" height="254" alt="image" src="https://github.com/user-attachments/assets/e4be3dd7-1685-4e1c-af79-ac379ff688fa" />

The tooltips also look fine now, the margins are symmetric on all sides as they should be :+1:
